funsor/ops/op.py

```diff
@@ -4,7 +4,25 @@
 from multipledispatch import Dispatcher
 
 
+class CachedOpMeta(type):
+    """
+    Metaclass for caching op instance construction.
+    """
+    def __call__(cls, *args):
+        try:
+            return cls._cache[args]
+        except KeyError:
+            instance = super(CachedOpMeta, cls).__call__(*args)
+            cls._cache[args] = instance
+            return instance
+
+
 class Op(Dispatcher):
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._cache = {}
+
```

Comment:
nit: I think you'll want to split this up into
```py
class CachedOp(Op, metaclass=CachedOpMeta):
    def __init_subclass__(cls, **kwargs):
        ...

class CachedOp(metaclass=CachedOpMeta):
    ...

class GetitemOp(CachedOp):  # Yay, no need for metaclass!
    ...
```

Reply:
The reason I didn't do this was that I wanted to keep caching out of the `Op` class hierarchy initially.  I also suspect we will want most ops to be cached in the future as an extra layer of robustness in cons-hashing and serialization, even though most common op constructors should only ever be called once anyway.